File: crates/krishiv/examples/stream_multi_source.rs
//! Real-time sliding window with Multi-Source Watermark synchronization.
//! Tested locally using in-memory streams without requiring a Kafka broker.
//! Run with: `cargo run -p krishiv --example stream_multi_source`

#![forbid(unsafe_code)]

use std::error::Error;
use std::sync::Arc;

use arrow::array::{Int64Array, StringArray};
use krishiv::{
    DataType, ExecutionMode, Field, MultiSourceWatermarkSpec, QueryResult, RecordBatch, Schema,
    Session, StreamBatch, WatermarkSpec,
};

fn main() -> Result<(), Box<dyn Error>> {
    let session = Session::builder()
        .with_execution_mode(ExecutionMode::Embedded)
        .build()?;

    let schema = Arc::new(Schema::new(vec![
        Field::new("timestamp", DataType::Int64, false),
        Field::new("device_id", DataType::Utf8, false),
    ]));
    let batch = RecordBatch::try_new(
        schema.clone(),
        vec![
            Arc::new(Int64Array::from(vec![1000, 2000, 3000, 8000])),
            Arc::new(StringArray::from(vec![
                "device-1", "device-2", "device-1", "device-2",
            ])),
        ],
    )?;

    let stream = session
        .memory_stream("sensor_stream", vec![StreamBatch::new(0, batch)])
        .unwrap();

    // Reconcile watermarks dynamically across both device identifiers
    let multi_source_wm = MultiSourceWatermarkSpec::new()
        .with_source_id_column("device_id")
        .source("device-1", WatermarkSpec::fixed_lag_ms(1000))
        .source("device-2", WatermarkSpec::fixed_lag_ms(2000));

    let windowed = stream
        .key_by("device_id")
        .with_event_time("timestamp")
        .with_multi_source_watermark(multi_source_wm)
        .sliding_window(10000, 5000); // 10s window size, sliding every 5s

    let results = windowed.collect()?;

    let batches = results.into_iter().map(|b| b.batch().clone()).collect();
    println!("{}", QueryResult::new(batches).pretty()?);

    Ok(())
}

File: crates/krishiv-python/examples/stream_multi_source.py
#!/usr/bin/env python3
"""Real-time sliding window with multi-source watermark synchronization in Python."""

import pyarrow as pa
import krishiv as ks

def main():
    # 1. Build session
    session = ks.Session.from_env()

    # 2. Prepare multi-device sensor events
    schema = pa.schema([
        ("timestamp", pa.int64()),
        ("device_id", pa.string()),
    ])
    batch = pa.RecordBatch.from_pydict({
        "timestamp": [1000, 2000, 3000, 8000],
        "device_id": ["device-1", "device-2", "device-1", "device-2"],
    }, schema=schema)

    # 3. Register as a bounded streaming relation (returns PyRelation with full pipeline API)
    relation = session.from_bounded_stream(
        "sensor_stream",
        [ks.Batch(batch)],
        watermark_column="timestamp",
        max_lateness_ms=1000,
    )

    # 4. Apply per-source watermarks and sliding window
    windowed = (
        relation
        .with_source_id_column("device_id")   # column that identifies each source
        .with_source_watermark("device-1", 1000)
        .with_source_watermark("device-2", 2000)
        .key_by("device_id")
        .sliding_window(10000, 5000)  # 10s window, 5s slide
    )

    # 5. Collect and print
    results = windowed.collect()
    pa_batches = [pa.record_batch(b.to_arrow()) for b in results]
    table = pa.Table.from_batches(pa_batches)
    print(table)

if __name__ == "__main__":
    main()
